lib/posix/options/fs.c

```diff
@@ -438,3 +438,23 @@ int ftruncate(int fd, off_t length)
 
 	return 0;
 }
+
+/**
+ * @brief Remove a directory.
+ *
+ * See IEEE 1003.1
+ */
+int rmdir(const char *path)
+{
+	return unlink(path);
+}
+
+/**
+ * @brief Remove a file.
+ *
+ * See IEEE 1003.1
+ */
+int remove(const char *path)
```

Comment:
This one is a bit tricky because remove has been a part of ISO C since 1989.

That means, this function should probably go into the common libc.

https://en.cppreference.com/w/c/io/remove

And in order to eliminate dependency cycles at the API level, the libc version of this should call the Zephyr `fs_unlink()` function directly rather than the POSIX `unlink()`.

If `CONFIG_FS` is not set, then `remove()` can return -1 and set errno to `ENOTSUP`.

With that, `rmdir()` and `remove()` still satisfy all of the requirements from POSIX (i.e. that they are effectively both calls to the same `fs_unlink()`.

Thoughts, @de-nordic ?

Reply:
I agree with @cfriedt  that we should rather directly call to system specific function, so ‘fs_unlink’ should be called.

As far as I understand https://pubs.opengroup.org/onlinepubs/9699919799/functions/remove.html#tag_16_485 it is part of both, I don’t mind having it implemented as part of POSIX; we can discuss further the right place, but at least we will have it, till we agree.

Shouldn’t we return ‘-ENOSYS’ https://docs.zephyrproject.org/latest/develop/api/design_guidelines.html#return-codes.

Reply:
@cfriedt  I can remove the `remove` api as it implement in the libc.

Reply:
It isn't implemented (yet) in the common libc, so I would suggest moving it there.

Reply:
@cfriedt  Are you Pointing to the zephyr native libc or something.  I think `remove` is implementated in the picolibc`modules/lib/picolibc/./newlib/libc/stdio/remove.c`.

Reply:
> @cfriedt  Are you Pointing to the zephyr native libc or something.

In `lib/libc/common`

> I think `remove` is implementated in the picolibc`modules/lib/picolibc/newlib/libc/stdio/remove.c`.

```cpp
int
remove (
       const char *filename)
{
  if (unlink (filename) == -1)
    return -1;

  return 0;
}
```

And what happens when `unlink()` is unavailable?

This is one of the reasons why we have `lib/libc/common` in Zephyr.

Since `remove()` is part of ISO C, and since POSIX is not required for ISO C applications, we need to implement `remove()` via Zephyr's native APIs (i.e. `fs_unlink()`). With that both POSIX and C are share a mutual dependency instead of a dependency cycle.